ISLANDORA-2205: Re-design `creative_commons` element

d-r-p opened the pull request:
Also, implement ISLANDORA-2208 and solve issues in #251

**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2205

* https://jira.duraspace.org/browse/ISLANDORA-2208
* #251

# What does this Pull Request do?

Re-design `creative_commons` element to solve bugs and implement new features. In particular, users can set default license in form builder, choose Public Domain licenses, override license versions and work without JS.

# What's new?

* Honour '#default_value'
* Allow CC0 & Public Domain Mark licenses
* Graceful non-JS degradation
* Allow overriding license version
* Allow setting license URI manually
* Validate versions, resp. license URIs
* Cache last query to API for speed-up
* Fix `<html>` wrapper of API response
* Renounce to value callback; populate values in element creation function
* Remove '#tree' logic in favour of manipulating `$form_state` directly

# How should this be tested?

* On a test VM, try to reproduce all the issues listed in the JIRA tickets and the above-mentioned PR.
* Pull this PR
* Verify that issues have disappeared (and no new ones arose).

# Additional Notes:

Since this PR basically re-designs the `creative_commons` element (in particular dropping the '#tree' manipulation), it could well be that
1. This code is more buggy than the previous one
2. There are neater ways of doing what is done here

A thorough code review is, therefore, indispensable.

Example:
* Does this change require documentation to be updated? **perhaps?**
* Does this change add any new dependencies?  **no**
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? **probably not, as functions only got extended by optional arguments**
* Could this change impact execution of existing code? **hopefully not**

# Interested parties
@adam-vessey @bondjimbond @rosiel @whikloj @DiegoPino @Islandora/7-x-1-x-committers

d-r-p committed:
ISLANDORA-2205: Re-design `creative_commons` element

Also, implement ISLANDORA-2208 and solve issues in
Islandora/islandora_xml_forms#251

Overview:
- Honor '#default_value'
- Allow CC0 & Public Domain Mark licenses
- Graceful non-JS degradation
- Allow overriding license version
- Allow setting license URI manually
- Validate versions, resp. license URIs
- Cache last query to API for speed-up
- Fix `<html>` wrapper of API response
- Renounce to value callback; populate values
  in element creation function
- Remove '#tree' logic in favor of manipulating
  `$form_state` directly

d-r-p commented on elements/includes/creative_commons.inc:
```diff
@@ -203,59 +182,505 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+            // Warn that the module default seems to be deprecated.
+            watchdog('xml_form_elements', 'Creative Commons default license seems to be deprecated. Please announce the developers.', array(), WATCHDOG_WARNING);
+          }
+          $current_query = xml_form_elements_creative_commons_make_value_array('', '', '', '', '', $license_uri);
```
Sorry for Travis' hiccup. I will mend the "[creative_commons.inc#L442-464](https://github.com/d-r-p/islandora_xml_forms/blob/1e821c0fe7078207d6344aaa1bf1b101cf868510/elements/includes/creative_commons.inc#L442-L464) and [creative_commons.inc#L489-511](https://github.com/d-r-p/islandora_xml_forms/blob/1e821c0fe7078207d6344aaa1bf1b101cf868510/elements/includes/creative_commons.inc#L489-L511) are duplicates" as soon as I get to it...

DiegoPino commented:
Hi, this all reads amazingly. I will have to discuss with further with you
and other committers since this fixes actual bugs (so goes into the
release) but improvements do not! So, this is a pretty large pull that
probably needs to be handled in a special way.
Will start by testing and we can go from there. This all is very good.
Thanks a lot

Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

d-r-p committed:
Code sanitation and minor changes

- Move API response storage code in separate functions
  to avoid code duplication
- Remove unnecessary '#weight' field in button element
- Move comment inside clause
- Update variables before delivering value, based on
  computed URI
- Avoid "Illegal choice in License Jurisdiction element."
  error due to empty  (set to 'international'
  by default)

d-r-p committed:
Minor change

If version was set to blank to retrieve newest license,
do not update that variable when not in manual mode.

bondjimbond commented:
Thanks, @d-r-p - this looks promising.

Hit an error on my first test: 
- Set default value to https://creativecommons.org/licenses/by-nc/4.0/
- Select "Creative commons attribution"
- Error appears: "Invalid CC license version (leave empty for newest)."
- Under the "Override license version" field, it auto-fills "creativecommons.org"

Something wrong with the "Override license version" field?

bondjimbond commented:
The aforementioned error does not show up if there is no "default value" entered.

bondjimbond commented:
Another update: The error occurs because I was using an HTTPS URI. Change to an HTTP URI, and the error does not occur.

Creative Commons website is HTTPS-only nowadays, so an HTTPS URI seems like a valid selection (or at least a reasonable mistake to expect). Can you modify so that HTTPS version is acceptable?

d-r-p commented:
Thank you, @bondjimbond, for testing!

Indeed, I am currently working on catching errors in `'#default_value'`, as well - I had overlooked this pitfall.

I totally agree with you, that `https` should be acceptable, but it is **NOT** (or so the API says).

Please allow some hours to push better error handling.

bondjimbond commented:
@d-r-p If the API dislikes https, perhaps one option would be to look for that in the default value field and warn the user? I don't suppose there's a way to verify the validity of the default value within Form Builder when configuring the element..

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -266,57 +701,132 @@ function xml_form_elements_creative_commons_ajax(&$form, &$form_state) {
+    ) : array()
+  ));
+  if ($response->code != 200) {
+    error_log("xml_form responded {$response->code}");
```
This is legacy code and needs to be replaced by a watchlog entry or directly by an inline (form) notification. Also, could be good to be sure if creative commons does not reply also with cached or redirects? Not mandatory, but better to be safe.

d-r-p replied:
@DiegoPino I will replace this with a watchdog entry.

Regarding redirect replies, we have three to go per default (see https://api.drupal.org/api/drupal/includes!common.inc/function/drupal_http_request/7.x). By that time we should hit `200`, no?

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -266,57 +701,132 @@ function xml_form_elements_creative_commons_ajax(&$form, &$form_state) {
-    }
-    return simplexml_load_string($response->data, 'SimpleXMLElement');
+  elseif ($version != '') {
+    return xml_form_elements_get_creative_commons_from_uri(xml_form_elements_creative_commons_value($commercial, $derivatives, $jurisdiction, $disabled, $version, $license));
```
This nested call makes reading and debugging more complicated. Also i see a lot of xml_form_elements_get_creative_commons_somethings here and this from _uri seems to be the first that actually says correctly where it is getting the value from but others like xml_form_elements_creative_commons_get_api_response's name are confusing because they do not respond with that only. Could it be that because of legacy code there is too much nesting?

d-r-p replied:
@DiegoPino Mmmhhh... I am not sure if I know exactly how to solve this nesting issue, but I will give it some thought (or do you have suggestions?).

The function names are, indeed, based on the legacy names (also the order of parameters) in order not to break people's code. Regarding the _misnomer_, would it be o.k. to call the function `xml_form_elements_creative_commons_get_api_response_from_cache()`, even though it might perform an API request, or does the problem you are seeing go deeper? (Also, I do not quite understand what _else_ they are responding, besides `FALSE` on failure - or did I miss your point?)

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -15,14 +23,27 @@
 function xml_form_elements_creative_commons($element, &$form_state) {
-  $creative_commons_base_url = "http://creativecommons.org/licenses/";
+  $license_options = array(
+    'none' => t('Do not output any license'),
```
Should this come from/match or be cached from https://api.creativecommons.org/docs/readme_15.html#locale-xx

d-r-p replied:
@DiegoPino I was thinking to this, initially, so that the choice is always up-to-date. But:
- do we also want to support the deprecated 'sample+' license? (we could nuke it, though)
- 'zero' and 'publicdomain' are identical (so we need to identify them)
- new licenses might break other logic (in particular checks)
- need default set, anyway, when API is unreachable
- I wanted to reduce API requests

I, then, decided against it, but I would be happy to implement this retrieval, if that is what is wanted.

DiegoPino reviewed (changes requested):
Minor change request, some need to avoid so much nesting (complicates following the code) and some questions about if hardcoding values v/s using the API. Or at least matching what the API gives you right now. Also about the defaults, We can check the default value against a regular expression if needed and give people some alert, but we need to be assuming that XML form builder is an advanced tool and there are and will be a lot of things you can put there and break the whole form. So a warning of some type is as far as i would go.

d-r-p commented:
@DiegoPino When you say warning, do you mean at form builder set-up time? In this case, a hint on how to do that would be much appreciated.

What I have been working on since yesterday is just the warning at form submission time (and, perhaps, telling the user to inform the admin if `'#default_value'` is problematic - I still have a hiccup there, though, hence the delay). Will that do, or do we need more?

DiegoPino commented:
@d-r-p Ok, the issue I see with default value validation at submit time is that sometimes users are tangential to islandora (means they are not even aware of the complexities) specially when using forms I fear some users won't know what/how and why they are getting that notice or what a default value is in this context, or worst, what can they do other than report it to someone and turn the computer off. Maybe I'm underestimating that type of user, but I was feeling the responsible for the form creation should be notified. That said, you approach is basically the right one.  If the admin creating the form gives it a test run it will see the problem pop up with your check. 👍

d-r-p committed:
Improve error handling, deconvolute logic, fix bugs

- Catch site setup errors (and issue warning for admin)
- Reduce level of nesting
- Add comments for explanation
- Rename and re-order utility functions
- Improve legacy code and deprecate legacy functions
- Fix minor bugs

d-r-p commented:
@DiegoPino So I tried to re-factor the whole thing and I hope it is now easier to understand (well, it cannot work without _some_ level of complexity, right?). Since I did not receive any replies to my questions/remarks within your code review, I stuck with some decisions I had originally made. Feel free to let me know if I should still change something (e.g. naming of functions, etc...).

@bondjimbond Could I ask you to give this another spin? Thank you so much!

DiegoPino commented:
Hi, sorry, I thought i had responded to the main question(s) in my post review message. The other questions you had seemed to me more like explanations of why you made the decisions but i probably read that wrong. I will test and re read all your code after breakfast. Thanks again for your work

bondjimbond commented:
@d-r-p It seems to be working nicely! Well done, and thanks for this.

I see that with the absence of a default value set in the element configuration, it now defaults to a manual URI. This is a change from the two prior defaults -- first CC-BY, then no license at all. 

I have no problem with this, just want to make sure it's (a) deliberate, and (b) generally agreed-upon as a good choice.

d-r-p committed:
"No license" selection on empty '#default_value'

d-r-p commented:
@bondjimbond Thank you very much! Indeed, you are right, this is neither a) deliberate, nor b) a good choice. Sorry that this escaped me. It should be o.k. now.

@bondjimbond @DiegoPino I found an issue, however (also present in the pre-#253 world):
When users choose to clean up their MODS from empty CC elements - as we suggest, actually - editing such objects will result in the `'#default_value'` being loaded rather than "no license". Should we fix/document this?

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -558,6 +558,11 @@ function xml_form_elements_creative_commons_get_values_from_uri($license_uri) {
 
+  // If no license URI is set, default to no output.
+  if (empty($license_uri)) {
+    $values['license_fieldset']['license_type'] = 'none';
```
nice

DiegoPino replied:
@d-r-p 
> When users choose to clean up their MODS from empty CC elements - as we suggest, actually - editing such objects will result in the '#default_value' being loaded rather than "no license". Should we fix/document this?

Not sure if that is an issue or how can that be fixed. Assuming the Administrator builds an XML form with a given default for CC, the default being loaded everytime the form is invoked without a given XML element seems expected. I mean the 'none' means NO license or means no XML element? If we define that well in docs and we don't break existing/expected functionality it should be good. I feel this could end being over-engineered and will eventually break anyway in the presence of a wrongly set :read :update xpath 😄

d-r-p replied:
@DiegoPino I have not tested this, but I have a feeling this is a general "issue" (i.e. pertaining to all form elements), anyway. Your comment seems to confirm this.

Regarding the _expected_ behaviour: I am not sure what to think. If the element remains empty, the `'#default_value'` will not be introduced. If the element is nuked, the element will populate with the `'#default_value'` upon edit. Since both cases stem from submitting the element with no value, I see an inconsistency for users. But, perhaps, I am overthinking this... In any case, this seems like a separate issue that can be discussed further elsewhere, right?

DiegoPino replied:
> Since both cases stem from submitting the element with no value, I see an inconsistency for users.

I agree

>  In any case, this seems like a separate issue that can be discussed further elsewhere, right?

Also true. We can only save one tree at a time and lately pulls are getting too big all over the world. Single pulls are easier to address and to review. So ++ on moving forward. Will do a review later today. Thanks again

d-r-p committed:
Tiny caching improvement

Avoid unnecessary second API query.

bondjimbond commented:
@DiegoPino Have the latest commits solved what you outlined in your review? If so I can give it a final test to confirm that all is well before merging.

DiegoPino commented:
@bondjimbond yes mostly including https://github.com/Islandora/islandora_xml_forms/pull/253/commits/ed8dbed1f5a6ddc8448ced04cbd3471f42215e72. I had tested everything and was about to give a review and then a new commit came (cache) so to be fully sure I will have to test again. Are you in a hurry to merge?

bondjimbond commented:
@DiegoPino Anxious and excited to make it happen, but not in a rush. :) Making the Creative Commons element work optimally is important to a project I'm working on - hence all the tickets about it lately.

DiegoPino commented:
Cool. Ok, I will release the block, but could you test it once again  (please) with the caching piece in place? Once this gets through, we will need a release pull too. Thanks!

bondjimbond commented:
@d-r-p @DiegoPino Got an error... possibly a minor one, but it's one that could crop up in regular life.

> Notice: Undefined index: license_uri in xml_form_elements_creative_commons_process_input() (line 712 of /var/www/drupal/sites/all/modules/islandora_xml_forms/elements/includes/creative_commons.inc).

To reproduce:
* Set a manual license URI
* Submit the ingest form without allowing the the "Selected License" image to load (click Submit without clicking off, or just click Submit too fast for the image to change)

If you do this, you get the error message, and the changes you made to that field are not recorded in the metadata.

d-r-p commented:
@DiegoPino Perhaps you should disapprove this PR again...

@bondjimbond Good catch!! I can confirm the issue. It actually goes further: whenever a sub-element of the `creative_commons` element changes and still has focus (including the drop-downs) and one submits, the value of that element is missing completely from the input array (`$form_state['input']`). You can actually reproduce part of this issue on the pre-#253 version of the form (changes not recorded - no error, however), but the current state of this is, of course, unacceptable.

The culprit seems to be AJAX (no trouble in the no-JS version, as far as I can tell) or, more precisely, the way I am using it (I guess I should not replace the whole fieldset, but rebuilding the whole form would be a shame). I have to investigate further...

DiegoPino commented:
Nah, i won’t disapprove, i just delegated testing to @bondjimbond <https://github.com/bondjimbond>. Should i test now?

Diego Pino Navarro
Metropolitan New York Library Council
599 11th Av. New York, NY 10036

> On May 7, 2018, at 4:14 PM, d-r-p <notifications@github.com> wrote:
> 
> @DiegoPino <https://github.com/DiegoPino> Perhaps you should disapprove this PR again...
> 
> @bondjimbond <https://github.com/bondjimbond> Good catch!! I can confirm the issue. It actually goes further: whenever a sub-element of the creative_commons element changes and still has focus (including the drop-downs) and one submits, the value of that element is missing completely from the input array ($form_state['input']). You can actually reproduce part of this issue on the pre-#253 <https://github.com/Islandora/islandora_xml_forms/pull/253> version of the form (changes not recorded - no error, however), but the current state of this is, of course, unacceptable.
> 
> The culprit seems to be AJAX (no trouble in the no-JS version, as far as I can tell) or, more precisely, the way I am using it (I guess I should not replace the whole fieldset, but rebuilding the whole form would be a shame). I have to investigate further...
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Islandora/islandora_xml_forms/pull/253#issuecomment-387189999>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AGn85xwJeHXpZgPp-6iYZlWA863amwZ8ks5twKsRgaJpZM4TrgtB>.
>

d-r-p commented:
To me, all this sounds extremely much like the issue described in https://www.drupal.org/project/drupal/issues/1736308. Indeed, the value of the field in focus does not even exist in the `$_POST` data.

Suggested workarounds seem to be to a) either store the old value of the field in a hidden field or b) to prevent form submission while AJAX is running. Should I try to come up with an attachable `.js`?

d-r-p committed:
Prevent form submission during AJAX requests.

bondjimbond commented:
@d-r-p @DiegoPino 

>To me, all this sounds extremely much like the issue described in https://www.drupal.org/project/drupal/issues/1736308. Indeed, the value of the field in focus does not even exist in the $_POST data.

If this is a Drupal bug, then it's not a bug with this PR and should be addressed separately, no? As-is, this is better than what we've got currently... But if your workarounds can work, awesome. I prefer an inelegant solution to (a) the bug or (b) not merging.

@DiegoPino Aside from the bug we're discussing now, everything else seems to work well. Please go ahead and do your tests.

DiegoPino commented:
@d-r-p i will test this today again. Not completely convinced about the ajax thing, i really need to test that further, also the issue reported there is for Drupal 8 and there are mixed opinions. Give me 24 hours please (sorry @bondjimbond )

d-r-p commented:
@DiegoPino Sure thing. Thank you for testing!

I was aware that the issue I mentioned is for Drupal8, but the description fits what is going wrong here.

Fact is that the value is missing completely from `$_POST`, and (to me) that seems to be due to some AJAX-related JS at work. Preventing submission in favour of letting AJAX finish has two benefits. First, the value is updated as intended by the user (issue pertaining also to pre-#253), and, second, the value will be present in `$_POST` (so as to prevent the error seen by @bondjimbond). If you have another idea on how to solve this (such as firing a `set_form_error` telling the user that AJAX did not finish), please share. For my part, I am wondering whether this should be solved for Drupal/AJAX in general (I am aware of code that prevents submission while files get uploaded via AJAX) - but that is, probably, a discussion to be had elsewhere...

One last thing: it would be awesome if @DiegoPino or @bondjimbond could quickly check the selectors I chose in [`elements/js/creative_commons.js`](https://github.com/Islandora/islandora_xml_forms/pull/253/commits/2fe9e6f4deab981c42855ee1f5339b339c88b997#diff-92c3e4b305630b1e5dfe106315954a3e) ([L9](https://github.com/Islandora/islandora_xml_forms/pull/253/commits/2fe9e6f4deab981c42855ee1f5339b339c88b997#diff-92c3e4b305630b1e5dfe106315954a3eR9), [L15](https://github.com/Islandora/islandora_xml_forms/pull/253/commits/2fe9e6f4deab981c42855ee1f5339b339c88b997#diff-92c3e4b305630b1e5dfe106315954a3eR15) and [L20](https://github.com/Islandora/islandora_xml_forms/pull/253/commits/2fe9e6f4deab981c42855ee1f5339b339c88b997#diff-92c3e4b305630b1e5dfe106315954a3eR20)) - I would not be surprised if there were better choices...

d-r-p commented:
@DiegoPino @bondjimbond Okay, I tested the alternative to disabling submission I had alluded to, i.e. issue a form error when input is missing. It is not very elegant and the user input gets lost, but something along the lines of

> "Manually enter a license URI" unavailable because the AJAX request did not complete. This means that your input got lost. Please click outside the Creative Commons form element "License" and wait until the AJAX request has finished before submitting the form.

could be displayed when the issue arises. Do you want me to clean up my testing code, wipe the `.js` part and submit a commit?

DiegoPino commented:
@d-r-p before making more changes let me debug this ajax thing. I had my fair share of ajax while coding for islandora webform module and I remember I dealt with it in some way different... but as said lets test this as it is and then we can go from there. I would prefer to avoid moving things around until fully tested. Thanks a lot

DiegoPino commented:
@d-r-p i'm still on this. Will report back in an hour or so

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+    watchdog('xml_form_elements', 'The Creative Commons API responded with code %code.', array('%code' => $response->code), WATCHDOG_WARNING);
+    return FALSE;
+  }
+  return simplexml_load_string($response->data, 'SimpleXMLElement');
```
Not sure about your simplexml return. You are assuming that stuff comming in $response->data is well formed and actually there. Nature of Internet tells me that a 200 code means just that the response is Ok, not that the context. http://php.net/manual/en/function.simplexml-load-string.php will evaluate to two different types of "FALSE".

d-r-p replied:
@DiegoPino You are right, we are (and were) a bit on shaky ground here. I left it for the moment. `FALSE` is fine. If you think I should catch more errors, let me know...

DiegoPino replied:
Ok, seems enough, in case of bad XML it will throw some warnings DUNO, if we should  just let that happen, probably ok

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+      $response = xml_form_elements_creative_commons_query_api_details($values['license_fieldset']['license_uri']);
+    }
+    // Cache response.
+    xml_form_elements_creative_commons_store_api_response($element_name, $storage_folder, $values, $response, $form_state);
```
You are caching 'False' too?

DiegoPino replied:
The cache piece makes read-ability complex. Adding a full function to store the cache but having the logic that reads from cache inline in this functions confuses me. I suggest removing cache totally or implementing read/write it in a simpler to reuse way. Also, as i stated in another comment, the cache gets almost never hit.

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+    $form_state['storage']['xml_form_elements'][$element_name][$storage_folder] = array();
+  }
+  $form_state['storage']['xml_form_elements'][$element_name][$storage_folder]['query'] = $values;
+  $form_state['storage']['xml_form_elements'][$element_name][$storage_folder]['response'] = ($response !== FALSE) ? $response->asXML() : FALSE;
```
edge case but still related to how the response is dealt with: $response !== FALSE is api failed or XML comming from API call has XML errors, but $response != FALSE if xml empty .

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+ * @param array $form_state
+ *   The form state.
+ */
+function xml_form_elements_creative_commons_store_api_response($element_name, $storage_folder, $values, $response, &$form_state) {
```
Why an extra hierarchy $storage_folder = 'api_data'? You are not keeping a cache per query, just a single per element name, so if people change arguments previous cache serves no good. And if they don't change, the same story because the function won't be triggered. What are the chances that the cache is actually reused with this implementation?

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+ *   The response from the REST API as a SimpleXMLElement, or FALSE if request
+ *   failed.
+ *
+ * Note that there may be a bug in the CC API. Querying, e.g., 'fi'
```
this warning this needs to go into documentation too. No way we can expect XML form users to read PHP code...

d-r-p replied:
@DiegoPino Excellent point! Where in the doc should it go? WIKI/Readme/...? If outside this repo, could you take care of that, please?

DiegoPino replied:
I would say in the XML forms wiki , github, and a note in the README.md (known issues?) @rosiel ?

d-r-p replied:
@DiegoPino @rosiel If you want, I can write a note (maybe an item in a new sub-section `### Known Issues` under `## Troubleshooting/Issues`?) in the `README.md`. On the other hand, this is actually outside the scope of this particular PR, so it could also be done later (in case we are tight on time before release)...

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+          'jurisdiction' => $values['license_fieldset']['license_jurisdiction'],
+        );
+      }
+      $response = xml_form_elements_creative_commons_query_api_get($api_path, $api_options);
```
`xml_form_elements_creative_commons_query_api_get` and `xml_form_elements_creative_commons_query_api_details` do exactly the same and look almost exactly the same, first allows each option to pass individually and second the full URI. So, why not use a single one? Less code. This pull is just huge.

d-r-p replied:
@DiegoPino True, so I moved the common code in a helper function. I left both functions since I though they could be useful on their own. If not, I can reduce to one...

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+    isset($form_state['storage']['xml_form_elements'][$element_name][$storage_folder]['response']) &&
+    ($form_state['storage']['xml_form_elements'][$element_name][$storage_folder]['response'] !== FALSE)) {
+    // Get response from storage.
+    $response = simplexml_load_string($form_state['storage']['xml_form_elements'][$element_name][$storage_folder]['response'], 'SimpleXMLElement');
```
Too much logic for too little gain. Cache seems non practical

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+  // Get current values, process them and store them in the element.
+  $input = isset($form_state['input'][$element['#name']]) ? $form_state['input'][$element['#name']] : NULL;
+  $values = xml_form_elements_creative_commons_process_input($element, $input, $form_state);
+  $element['#value'] = $values;
```
Does this mean, that whatever the user does, $element's value will be $value?
I guess you all tested this pretty well, but i was in the impression that input was not to be trusted and to allow multi step to work correctly in a custom element like this you needed to make sure that form_set_value() is called if the value submitted by the users needs to be changed/something to make sure it  persists.  https://api.drupal.org/api/drupal/includes%21form.inc/function/form_set_value/7.x

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+ * function. This has the side effect that users cannot go to the previous step
+ * in a mulit-step form before resolving erroneous choices.
+ */
+function xml_form_elements_creative_commons_process_input(&$element, $input, &$form_state) {
```
WHy does xml_form_elements_creative_commons_process_input require $input to be passed if $input comes from &$element and can be fetched from there?

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,941 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+  $element['license_fieldset']['license_type'] = array(
+    '#type' => 'select',
+    '#title' => t('Select a license type'),
+    '#description' => t('Select "@option" to avoid outputting a creative commons license. This will still output a blank element in the resulting XML, so make sure to use a cleanup template to remove it.', array('@option' => $license_options['none'])),
```
This could appear only if "none" is selected, drupal's ajax re-renders the element on background, so could accomplished. Not a big deal so only if possible. "This will still output a blank element in the resulting XML, so make sure to use a cleanup template to remove it" things like this make sense if the form user is the admin and understands dark logic of islandora. Same message will freak out any metadata user that is not part of this religion.

d-r-p replied:
Oops! I had misread the requirement. Now it should be o.k..

DiegoPino reviewed (dismissed):
@d-r-p @bondjimbond this pull is just too complicated to review :sad. I had a  few AJAX time outs and  ajax related errors during test and $form_state between steps gets crazy. I feel (mostly convinced) that this pull, trying to address all issues is over engineered (i mean well) and some things can be removed like the cache, that is never hit to at least make reviewing the code something a single committer can do. I know this is different that checking if well known use cases work.  Stuff like values validation should only happen (extreme here) when people provide their own URI, if not, i suppose we don't need to run regex on the resulting URI and the defaults too. Also, i see there is a lot of hacking with form values/states. '#input' and even '#value' key all over the place when we should deal with the state. The ajax really runs slow, i guess because of all the tests done. Not sure if I can review this further as fast as you need without A) asking for changes and a better test procedure, since 918 new lines and calls functions inside calls functions inside functions calls is huge and redundancy and deep nesting needs to go.
B) Agreement that you are happy with how this is and you can simply vote and forget my comments.  I will ask from some changes here (not everything needs to be addressed, but if not addressing my concerns, please at least explain the why). I fear that if this goes code will become untouchable. Thoughts?

DiegoPino commented:
@d-r-p @bondjimbond could you please bring this pull into today's release call? I really appreciate the work done here, but it went pretty into the unmanageable part once it reached 9xx lines of code. I would really like you can get another pair of eyes willing to do a sincere review and I would, of course, like to get this into the release. Maybe I'm not the one for this (sick day). @rosiel ping here. Could you make sure this one gets space during the call? Thanks, guys and sorry. Testing this beast is not easy.

bondjimbond commented:
@DiegoPino Lots of comments there! I'm at a conference today so won't be in the release call, sadly. Will @d-r-p be there?

DiegoPino commented:
@d-r-p i just found another issue. If you toggle between none and CC0 and then enter manually, the manual field gets filled with http://creativecommons.org/publicdomain/zero/1.0/ If you change later to  CCA, version gets set to 1.0, which won't work for most of the countries I selected. But since the user never added 1.0, it can confuse interactions. Means, CC0 should only set URI, not a version. Does this make sense? Also if I add 4.0 as the desired version, the version field gets cleared out, but the license is correctly fetched as V4., but with 3.0 it stays in 3.0. I get the why, but I feel the logic is complex to follow. I get that 4.0 is right now the "latest" but if I fill 4.0 I would want 4.0 to stay there for consistency?. Finally (sorry), it is impossible to make the form required. Since "don't output" is a valid option... whatever you put, you can submit (almost, only if not errored) but that is not the same as required value.

rosiel commented:
(discussed on the release call): It sounds like there are outstanding issues with this, and it'd be great to get more eyes on it.  We know it solves bugs, but also introduces a lot of new stuff. We want to make sure we don't 'destabilize' XML form builder for this release, and we think this is gonna need more time.

DiegoPino commented:
@d-r-p ping. Need a call/mail to discuss this? Let me know if i can help

d-r-p committed:
Address (most) comments by @DiegoPino

- Add checkbox for version overriding in favour of not
    clearing version field for newest licenses
- Support '#required'
- Refrain from using `$form_state['input']` entirely
- Remove "cache" (and related functions), but use temporary
    storage to prevent re-requesting the API
- Move common code of the two API callers into helper
    function
- Issue `form_set_value()`
- Remove unnecessary `$input` parameter in
    `xml_form_elements_creative_commons_get_processed_values()`
- Hide technical warning when no license output is selected
- Improve behaviour for manually entered URIs (in particular,
    if a known license is entered with up-to-date version,
    unset version overriding)

DiegoPino commented:
@d-r-p WOW! A lot of good stuff there 👍  Stay tuned, will review as soon as I can.

d-r-p commented:
@DiegoPino @bondjimbond I (hopefully) addressed all the quirks. Since I moved away from using `$form_state['input']` altogether, I had to re-design the whole thing, again (it took me _ages_ to figure out how to make it work when in multi-step forms and going back). Sorry @DiegoPino for _even more_ code to review now (but addressing your comments/wishes needed some additional lines, too).

The upshot: please re-test as thoroughly as before. I really hope no new bugs crept in, but you never know... Thank you so much guys!!

d-r-p commented:
Oops! Seems I need to re-write a few lines for older PHP versions. Apologies.

d-r-p committed:
Remove PHP 5.5+ syntax

d-r-p committed:
Improve error message on unknown jurisdiction in manual mode

Also, improve consistency for newly introduced temporary variable.

d-r-p committed:
Mend the license type description for none vs. some license

bondjimbond commented:
@d-r-p You are a hero. I just tested this, and every bug/issue we'd found has been fixed. I quite like the solution to the Ajax issue: No longer lets you click the "Update" button until Ajax has finished running. Elegant. And we've got some simple validation for the URI (warning that HTTPS is no good, must end with a slash).

@DiegoPino This looks good to go in my eyes. Was there anything left from your review still to address?

bondjimbond reviewed (approved):
Looks good to me. No more bugs that I can find.

rosiel commented:
WRONG PULL REQUEST!! I did NOT mean to dismiss @DiegoPino's review on this PR.(sorry, i confused this PR with https://github.com/Islandora/islandora_badges/pull/19) 

Please wait on him to confirm whether his concerns are outstanding.

DiegoPino commented:
Guys, have been traveling/ conference this week sorry and on an airpory
now. Will check the code tomorrow (yey back home!!) i’m pretty sure @d-r-p
would be extra angry at me if i approved without having tested it at least
twice!
Thanks!!!

El El jue, 7 de jun. de 2018 a las 15:01, Rosemary Le Faive <
notifications@github.com> escribió:

> WRONG PULL REQUEST!! I did NOT mean to dismiss @DiegoPino
> <https://github.com/DiegoPino>'s review on this PR.(sorry, i confused
> this PR with Islandora/islandora_badges#19
> <https://github.com/Islandora/islandora_badges/pull/19>)
>
> Please wait on him to confirm whether his concerns are outstanding.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora/islandora_xml_forms/pull/253#issuecomment-395563505>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn856edoqWInNExcM8sWrAlvV4jPXTUks5t6ZSOgaJpZM4TrgtB>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

d-r-p commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,1089 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+  if (empty($license_uri) || (
+    preg_match("#^" . XML_FORM_ELEMENTS_CREATIVE_COMMONS_BASE_URL . "[a-z]+/[a-z+-]+/[0-9]\.[0-9]/([a-z]+/)?$#", $license_uri) &&
+    (!preg_match("#/by[-/]#", $license_uri) ||
+      preg_match("#^" . XML_FORM_ELEMENTS_CREATIVE_COMMONS_BASE_URL . xml_form_elements_creative_commons_get_license_path('by') . "/by((-nc)?(-nd|-sa?)|(-nd|-sa)?(-nc)?)/#", $license_uri))
```
Sorry... Just realised that this line should read (switch `?` and `)` in columns 149-150):
```php
      preg_match("#^" . XML_FORM_ELEMENTS_CREATIVE_COMMONS_BASE_URL . xml_form_elements_creative_commons_get_license_path('by') . "/by((-nc)?(-nd|-sa)?|(-nd|-sa)?(-nc)?)/#", $license_uri))
```

DiegoPino commented:
Ok. @d-r-p Let me know when you are done with your self review, thanks for
letting us know, and Brandon, no rush on merging until reviewed, even when
UI manual testing passes (good) code logic as secondary but definig pass is
always needed



El El jue, 7 de jun. de 2018 a las 17:34, d-r-p <notifications@github.com>
escribió:

> *@d-r-p* commented on this pull request.
> ------------------------------
>
> In elements/includes/creative_commons.inc
> <https://github.com/Islandora/islandora_xml_forms/pull/253#discussion_r193917911>
> :
>
> > +}
> +
> +/**
> + * Test the Creative Commons license URI format validity.
> + *
> + * @param string $license_uri
> + *   The license URI.
> + *
> + * @return bool
> + *   TRUE if format is valid, FALSE otherwise.
> + */
> +function xml_form_elements_creative_commons_license_uri_has_valid_format($license_uri) {
> +  if (empty($license_uri) || (
> +    preg_match("#^" . XML_FORM_ELEMENTS_CREATIVE_COMMONS_BASE_URL . "[a-z]+/[a-z+-]+/[0-9]\.[0-9]/([a-z]+/)?$#", $license_uri) &&
> +    (!preg_match("#/by[-/]#", $license_uri) ||
> +      preg_match("#^" . XML_FORM_ELEMENTS_CREATIVE_COMMONS_BASE_URL . xml_form_elements_creative_commons_get_license_path('by') . "/by((-nc)?(-nd|-sa?)|(-nd|-sa)?(-nc)?)/#", $license_uri))
>
> Sorry... Just realised that this line should read (switch ? and ) in
> columns 149-150):
>
>       preg_match("#^" . XML_FORM_ELEMENTS_CREATIVE_COMMONS_BASE_URL . xml_form_elements_creative_commons_get_license_path('by') . "/by((-nc)?(-nd|-sa)?|(-nd|-sa)?(-nc)?)/#", $license_uri))
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora/islandora_xml_forms/pull/253#pullrequestreview-126991740>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn85xn8q9mEnn9nhaacq1JGEuIY8nRKks5t6bh-gaJpZM4TrgtB>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

d-r-p committed:
Correct typo in regular expression

d-r-p commented:
@DiegoPino I caught that typo by accident. If TravisCI is happy, you're good to go for testing. Thank you!!

d-r-p commented:
@DiegoPino I think TravisCI timed out when trying to clone dependencies from GitHub. Perhaps you could re-start? Thanks!

DiegoPino commented:
@d-r-p travis is green. Will do the final review later today and we should have green light if no outstanding issue appears. Thanks for your work and your will to persue this further

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,1089 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+    '#limit_validation_errors' => array(array($element['#name'])),
+    '#submit' => array('xml_form_elements_creative_commons_license_update_submit'),
+    '#return_value' => 'update-license',
+    '#prefix' => '<noscript>',
```
lovely

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,1089 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+
+  // If there is no input, or the temporary storage contains data (meaning
+  // that the validator already exectued), populate the element with values.
+  if (!isset($form_state['input'][$element['#name']]) || isset($form_state['temporary'][$element['#name']])) {
```
good!

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,1089 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+  if (!isset($values['license_fieldset']) || !empty($missing_values['license_fieldset'])) {
+    $values['license_fieldset'] = array_merge($empty_values['license_fieldset'], isset($values['license_fieldset']) ? $values['license_fieldset'] : array());
+    drupal_set_message(t('An internal error occurred while processing the Creative Commons form element "@display_name" (unexpected data loss). Please inform the site administrator.', array("@display_name" => $element['license_fieldset']['#title'])), 'error');
+    watchdog('xml_form_elements', 'Unexpected data loss when processing element "@element_name" in "@form_name". Probably you just found a bug.', array('@element_name' => $element['#name'], '@form_name' => $form_state['storage'][FormStorage::STORAGE_ROOT]['name']), WATCHDOG_ERROR);
```
Just to be sure @d-r-p, did you test that `FormStorage`class is available in this scope?

d-r-p replied:
@DiegoPino well, I did nuke an array element to trigger the error for testing and it seemed to work just fine...

DiegoPino replied:
Can you confirm this is also the case in the form builder UI  as it is on a live XML form multi step ingest? Sorry for being overprotective...

d-r-p replied:
@DiegoPino I think I _did_ test this in both scenarios (form builder view form + ingest new object), but will re-test to be extra sure.

d-r-p replied:
@DiegoPino I tested scope availability by triggering watchdog in [L347](https://github.com/Islandora/islandora_xml_forms/pull/253/files/f6a0d63b322aa7613b20d266d219090c7cdbb4ef#diff-0fca5b132da440f97135da3015937c40R347). Both scenarios seem to have `FormStorage` available. Should the other two occurrences be tested separately, or are we good?

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,1089 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+            form_set_error($element['#name'] . '][license_fieldset][license_uri', t("CC license jurisdiction unknown or not applicable for this license."));
+          }
+          else {
+            form_set_error($element['#name'] . '][license_fieldset][license_uri', t("CC license URI seems to be invalid. Perhaps you would prefer to choose one automatically?"));
```
like the suggestion to let the logic handle it

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,1089 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+    }
+    else {
+      // We could not reach the API.
+      drupal_set_message(t("Could not reach Creative Commons API to validate your input."), 'warning');
```
This `else` here, what happens if Creative Commons cannot be reached? Do we allow a save anyway? Or do we trigger an $error = TRUE? To be sure we don't get bad data saved? Thoughts?

d-r-p replied:
@DiegoPino the behaviour thus far (and, therefore, maybe the expected behaviour) was to "fall back" on manually creating the URL when Creative Commons is unreachable. I would incline to keep this (thus allowing unvalidated URLs in this case), and let the user decide upon the warning what should be done. What are @bondjimbond's thoughts about that?

DiegoPino replied:
Ok, convinced... then lets tell people that if API fails manual is still an option? Too much to ask? Probably just a few more words after .. "validate your input" ...

d-r-p replied:
@DiegoPino If API is unavailable, users can still use _automatic_ mode. However the versions will be wrong (in all likelyhood). Or did I not understand your point?

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,1089 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+ *
+ * @return mixed
+ *   SimpleXMLElement the return from the REST API.
+ *   FALSE if failed request.
```
"if failed request "... or $disabled == disabled? Crazy logic

d-r-p replied:
@DiegoPino note that this is a legacy function that we are deprecating now. Do you want a modification in the documenting comment (and, btw, it would be "or `$disabled == TRUE`")?

DiegoPino replied:
legacy or not, it is marked as new lines added by this pull. Means you are contributing to this.

d-r-p replied:
@DiegoPino No worries, I can update that...

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -72,269 +124,1089 @@ function xml_form_elements_creative_commons($element, &$form_state) {
+ * @param bool $disabled
+ *   If TRUE, return empty string; otherwise generate the URI.
+ *
+ * @return string
```
or empty if disabled?

d-r-p replied:
@DiegoPino again, this is deprecated function henceforth. I figured the empty string `''` is still a `string` regarding type...

DiegoPino replied:
The value of the element == '' ?

DiegoPino replied:
If you are ok with that i'm ok

d-r-p replied:
@DiegoPino not sure I got your issue, then...

DiegoPino replied:
Just being picky, if disabled the function is not "really" returning the value of the element (line 1197)... forget it.

DiegoPino commented on elements/js/creative_commons.js:
```diff
@@ -0,0 +1,25 @@
+(function($) {
+  Drupal.behaviors.xmlFormElementCreativeCommons = {
+    attach: function (context, settings) {
+      $('[id|=license-fieldset]', context).parents('form').once(function(){
```
cool

DiegoPino commented:
@d-r-p so by law i'm only allowed to review one 1000 lines pull request a year! But hey, this looks good and required those lines. I left some open questions for you, minor about PHPdoc stuff and one about a class :: method that is being invoked that i'm not totally sure is in context defined there, since you have you dev environment there could you check? All code looks pretty awesome, had some trouble following the many if-elses somethings but much easier to follow than before! i wish we had more pulls (but shorter!) during releases, great!

DiegoPino commented on elements/includes/creative_commons.inc:
```diff
@@ -2,29 +2,79 @@
- *   The form state.
+ * Constants
+ */
+const XML_FORM_ELEMENTS_CREATIVE_COMMONS_BASE_URL = 'http://creativecommons.org/';
```
Last one here. Any reason to stay with the old HTTP instead of HTTPS?

d-r-p replied:
@DiegoPino If you remember [comment 385427248](https://github.com/Islandora/islandora_xml_forms/pull/253#issuecomment-385427248), the API considers `https` as invalid (and returns `http` URLs itself). This is unfortunate, as all the links can be reached via SSL, but we have to live with it.

DiegoPino replied:
I know but it scares me that in 2 years that will be not the case anymore... =(

d-r-p replied:
@DiegoPino true, which means we will have to modify then (and mend existing objects - which might actually be the reason for CC to vote against updating the URL thus far).

Perhaps the proper way to deal with this would be to _store_ the non-SSL URL, but to _render_ it as a `https` link.

DiegoPino replied:
@bondjimbond @rosiel what do you think? I kinda prefer storing good data long-term valid stuff (`https`) and validating against the `HTTP` one for now while doing a hot swap of protocols using `parse_url`or similar?. But can live with any style really.  Is there a proper style? Who said URIs where stable!!!!!! Damn WEB

d-r-p replied:
@DiegoPino :+1: for getting @bondjimbond's and @rosiel's opinion. FWIW, I think (gut feeling) they (CC) might be considering the license URI more like something similar to an XML namespace, rather than a web target. So they _might_ keep them stable after all...

bondjimbond replied:
My gut instinct is to stick with HTTP, since that is what the data currently requires, and we can't make assumptions about what CreativeCommons will and won't do. Is there any reason we would expect that HTTP URIs would no longer be valid at some point in the future, while still keeping the equivalent HTTPS URI valid?

If anything, wouldn't they just force redirect you to HTTPS when you visit the site? HTTP links should still be valid long-term as far as I understand.

DiegoPino replied:
Ok, lets do that then, lets saty on http. Is the CC API versioned? Can we say "this code works for version Blabla .0?"

d-r-p replied:
@DiegoPino It is! We are using here the (currently stable) 1.5 version (see https://api.creativecommons.org/docs/). Where would you like that info to be shown?

DiegoPino replied:
Yep, Ok, will just put that in the release notes. Should be enough right? README could be also a good option. Thanks!

d-r-p replied:
@DiegoPino Except for the newly introduced "Known Issues" subsubsection, there is no particular mention of the CC element in the README, right? So, perhaps, putting this info there would be a bit too specific (after all, it is the README of the _huge_ `islandora_xml_forms` module)? But I will not object if you or others insist in putting a note there (but where?)...

DiegoPino commented:
@d-r-p about the known issue and Doc. We are "kinda" trying to enforce coders to document also their exceptions. But if you all feel it is out of scope, to avoid this getting lost in space and time. a JIRA ticket after merge (to make sure we can mark the offending lines in the merged code v/s the to be merged code in the pull) is good for me too. Whatever is also less work. Or we can pass this to documentation IG? Or even as release notes..

d-r-p commented:
@DiegoPino
> We are "kinda" trying to enforce coders to document also their exceptions. 

:+1: for that! And I will gladly modify the documentation about the return values in the two deprecated functions.

Just for the record, though, those exceptions were missing altogether in the existing code doc, and - at least - I mention these exceptions already in the parameter documentation.

Regarding the _known issue_ with the CC API, let us wait on opinions from @rosiel. Or should I add a sub-section in `README.md` as suggested in [comment r194447782](https://github.com/Islandora/islandora_xml_forms/pull/253#discussion_r194447782) right away?

d-r-p commented:
@DiegoPino So, from what I see I should
- update the return value doc comment for the two deprecated functions
- add a "Known Issues" subsubsection in `README.md` with the single item about the CC API failure on special Finish license

Also, to recapitulate, the following questions are still open
- [comment r194463712](https://github.com/Islandora/islandora_xml_forms/pull/253#discussion_r194463712): is `FormStorage` availability sufficiently tested?
- [comment r194454426](https://github.com/Islandora/islandora_xml_forms/pull/253#discussion_r194454426): is there more to tell to the user when CC API is unavailable? Or should we even deal with it differently?
- [comment r194457109](https://github.com/Islandora/islandora_xml_forms/pull/253#discussion_r194457109): should we try to store `https` CC URLs? (waiting for opinions from @bondjimbond and @rosiel)

Did I forget anything?

DiegoPino commented:
@d-r-p you did not forget anything =)

d-r-p commented:
@DiegoPino what about the first two open questions? Anything I can deal with right away?

DiegoPino commented:
Nah. Just comments, nothing blocking. Will wait for Rosie to answer the
other piece and then we can maybe deal with that?

El El lun, 11 de jun. de 2018 a las 15:09, d-r-p <notifications@github.com>
escribió:

> @DiegoPino <https://github.com/DiegoPino> what about the first two open
> questions? Anything I can deal with right away?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora/islandora_xml_forms/pull/253#issuecomment-396353164>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn85yYvVZheoJqNe97Rxnsto-VR4ECfks5t7sCGgaJpZM4TrgtB>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

d-r-p commented:
@DiegoPino I was more referring to whether or not _my replies_ (see comments [r194463712](https://github.com/Islandora/islandora_xml_forms/pull/253#discussion_r194463712) and [r194454426](https://github.com/Islandora/islandora_xml_forms/pull/253#discussion_r194454426)) sufficiently cleared up the issues you raised...

Regarding the `README.md`, is the following acceptable?
> ### Known Issues
>
>* **Creative Commons Form Element**: Note that there seems to be a bug in the Creative Commons API triggered by certain combinations of _License Jurisdiction_ and other parameters. Querying, e.g., "Yes, as long as others share alike" for _Allow modifications of your work?_ and "Finland" for _License Jurisdiction_ results in a CC-BY-SA, resp. CC-BY-NC-SA, _4.0 International_ license, instead of the corresponding _1.0 Finland_ version (which exists and is not deprecated). The only way to obtain the correct license in this case is to enter `http://creativecommons.org/licenses/by-sa/1.0/fi/`, resp. `http://creativecommons.org/licenses/by-nc-sa/1.0/fi/`, _manually_ (choose "Manually select a license URI" for _Select a license type_). Also, this kind of licenses cannot be set by administrators in Form Builder as _Default Value_ for the Creative Commons Form Element.

as a subsubsection to "Troubleshooting/Issues"?

DiegoPino commented:
I thought i added some emojis to state “yes d-r-p, you addressed my
concerns “ maybe they did not stick? Anyway, i’m just waiting for Rosie now

El El lun, 11 de jun. de 2018 a las 15:40, d-r-p <notifications@github.com>
escribió:

> @DiegoPino <https://github.com/DiegoPino> I was more referring to whether
> or not *my replies* (see comments r194463712
> <https://github.com/Islandora/islandora_xml_forms/pull/253#discussion_r194463712>
> and r194454426
> <https://github.com/Islandora/islandora_xml_forms/pull/253#discussion_r194454426>)
> sufficiently cleared up the issues you raised...
>
> Regarding the README.md, is the following acceptable?
>
> Known Issues
>
>    - *Creative Commons Form Element*: Note that there seems to be a bug
>    in the Creative Commons API. Querying, e.g., "Yes, as long as others share
>    alike" for *Allow modifications of your work?* and "Finland" for *License
>    Jurisdiction* results in a CC-BY-SA, resp. CC-BY-NC-SA, *4.0
>    International* license, instead of the corresponding *1.0 Finland*
>    version (which exists and is not deprecated). The only way to obtain the
>    correct license in this case is to enter
>    http://creativecommons.org/licenses/by-sa/1.0/fi/, resp.
>    http://creativecommons.org/licenses/by-nc-sa/1.0/fi/, *manually*
>    (choose "Manually select a license URI" for *Select a license type*).
>    Also, this kind of licenses cannot be set as #default_value for the
>    element by administrators.
>
> as a subsubsection to "Troubleshooting/Issues"?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Islandora/islandora_xml_forms/pull/253#issuecomment-396361959>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGn855borgU1Ohh10Sa1OQNi58Q8Ybzrks5t7se5gaJpZM4TrgtB>
> .
>
-- 
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

d-r-p commented:
@DiegoPino no emojis in those comments that I can see (even after SHIFT-reload), but I am happy those two issues are resolved now for you...

Commit with the mentioned modifications coming up.

d-r-p committed:
Address @DiegoPino's code review

- Adjust return value code documentation for the two deprecated
    functions
- Add "Known Issues" sub-sub-section in `README.md` to document
    the CC API bug for Finish CC licenses

DiegoPino commented:
@bondjimbond @d-r-p this seems to be good. I will give @rosiel some more time to intervene or comment on the open questions and if nothing opposes will merge in 24 hrs. But, if @bondjimbond want´s to merge (since he tested also and code reviewed) he is in absolute magnanimous power to do so (but please wait for Rosie in case she wants to say something). Once this is in, can we get a release pull too? This is all good stuff. Thanks folks

bondjimbond commented:
Thanks for the very thorough review, @DiegoPino, and a hell of a complex bit of coding, @d-r-p. This is great stuff.

I don't mind who hits the merge button, so long as it's merged! Whoever gets there first in 21 hours is welcome to it. :)

Unfortunately we didn't quite break @bryjbrown's comment record of 139 (https://github.com/Islandora/islandora_scholar/pull/288), but we came close.

DiegoPino commented:
@d-r-p will squash merge now. If not the history will be a bit confusing. Again, my respects for some well-written code and a good/honest discussion.

d-r-p commented:
:tada: Thank you so much, @bondjimbond, for your repeated super-thorough testing and, @DiegoPino, for re-re-reviewing this (huge) PR until we got where we want to be! :tada:

It was a pleasure getting this done!

DiegoPino commented:
Our pleasure to have you around @d-r-p , making islandora a better place, community, and software. Thanks man
